src/sql/mod.rs

```diff
@@ -0,0 +1,849 @@
+            writer: RwLock::new(None),
+            reader: RwLock::new(None),
+        }
+    }
+}
+
+impl Drop for Sql {
+    fn drop(&mut self) {
+        async_std::task::block_on(self.close());
+    }
+}
+
+impl Sql {
+    pub fn new() -> Sql {
+        Self::default()
+    }
+
+    /// Checks if there is currently a connection to the underlying Sqlite database.
+    pub async fn is_open(&self) -> bool {
+        // in read only mode the writer does not exists
+        self.reader.read().await.is_some()
+    }
+
+    /// Closes all underlying Sqlite connections.
+    pub async fn close(&self) {
+        if let Some(sql) = self.writer.write().await.take() {
+            sql.close().await;
+        }
+        if let Some(sql) = self.reader.write().await.take() {
+            sql.close().await;
+        }
+    }
+
+    /// Opens the provided database and runs any necessary migrations.
+    /// If a database is already open, this will return an error.
+    pub async fn open<T: AsRef<Path>>(
+        &self,
+        context: &Context,
+        dbfile: T,
+        readonly: bool,
+    ) -> anyhow::Result<()> {
+        dbg!(dbfile.as_ref());
+        if self.is_open().await {
+            error!(
+                context,
+                "Cannot open, database \"{:?}\" already opened.",
+                dbfile.as_ref(),
+            );
+            return Err(Error::SqlAlreadyOpen.into());
+        }
+
+        // Open write pool
+        if !readonly {
+            let config = SqliteConnectOptions::new()
+                .journal_mode(SqliteJournalMode::Wal)
+                .filename(dbfile.as_ref())
+                .read_only(false)
+                .busy_timeout(Duration::from_secs(10))
+                .create_if_missing(true)
+                .synchronous(SqliteSynchronous::Normal);
+
+            let writer_pool = PoolOptions::<Sqlite>::new()
+                .max_connections(1)
+                .after_connect(|conn| {
+                    Box::pin(async move {
+                        let q = r#"
+PRAGMA secure_delete=on;
+PRAGMA temp_store=memory; -- Avoid SQLITE_IOERR_GETTEMPPATH errors on Android
+"#;
+
+                        conn.execute_many(sqlx::query(q))
+                            .collect::<std::result::Result<Vec<_>, _>>()
+                            .await?;
+                        Ok(())
+                    })
+                })
+                .connect_with(config)
+                .await?;
+            {
+                *self.writer.write().await = Some(writer_pool);
+            }
+        }
+
+        // Open read pool
+
+        let config = SqliteConnectOptions::new()
+            .journal_mode(SqliteJournalMode::Wal)
+            .filename(dbfile.as_ref())
+            .read_only(false)
+            .busy_timeout(Duration::from_secs(10))
+            .synchronous(SqliteSynchronous::Normal);
+
+        let reader_pool = PoolOptions::<Sqlite>::new()
+            .max_connections(num_cpus::get() as u32)
```

Comment:
Using number of CPUs doesn't really make sense here as we don't use threads. Even if machine is some cheap VM with a single core running a bot, we may need more than one connection if we spawn multiple async tasks. Some fixed number like 5 is more predictable in this case.

Reply:
The runtime uses threads under the hood, which means that we have at most ~ num_cpus threads trying to access the database, which is why I chosse this number.

Reply:
not sure, how `num_cpus` is actually related to the max. number of threads; also this would not really fit as there is also the ui-thread and the writer thread.

in practise, i tried that on iphone7 which has num_cpus=2, that felt much _less_ snappier than using num_cpus=10 which is the default for sqlx (and was also the default for rusqlite and was used in the past), i did a pr and commented a bit more at https://github.com/deltachat/deltachat-core-rust/pull/2324

Reply:
It's possible to have 1 CPU and multiple threads/tasks doing operations that start a `SELECT` statement and do network operations on each of it's result in parallel. Having only one read connection will make them block each other. So number of CPUs should not limit them in this case.